dbms/src/IO/Compression/CompressionCodecFactory.cpp

```diff
@@ -0,0 +1,214 @@
+    if (setting.method == CompressionMethod::LZ4HC)
+        return getStaticCodec<CompressionCodecLZ4HC>(setting);
+    if (setting.method_byte == CompressionMethodByte::LZ4)
+        return getStaticCodec<CompressionCodecLZ4>(setting);
+    if (setting.method_byte == CompressionMethodByte::ZSTD)
+        return getStaticCodec<CompressionCodecZSTD>(setting);
+    if (setting.method_byte == CompressionMethodByte::NONE)
+        return getStaticCodec<CompressionCodecNone>(setting);
+
+#if USE_QPL
+    if (setting.method_byte == CompressionMethodByte::QPL)
+        return getStaticCodec<CompressionCodecDeflateQpl>(setting);
+#endif
+
+    if constexpr (!IS_DECOMPRESS)
+    {
+        // If method_byte is Lightweight, use LZ4 codec for non-integral types
+        // If method_byte is DeltaFOR/RunLength/FOR, since we do not support use these methods independently, so there must be another codec, use it directly.
+        if (!isInteger(setting.data_type) && setting.method_byte == CompressionMethodByte::Lightweight)
+            return getStaticCodec<CompressionCodecLZ4>(setting);
+        else if (!isInteger(setting.data_type))
+            return nullptr;
+    }
+
+    switch (setting.method_byte)
+    {
+    case CompressionMethodByte::Lightweight:
+        return std::make_unique<CompressionCodecLightweight>(setting.data_type);
+    case CompressionMethodByte::DeltaFOR:
+        return getStaticCodec<CompressionCodecDeltaFOR>(setting);
+    case CompressionMethodByte::RunLength:
+        return getStaticCodec<CompressionCodecRunLength>(setting);
+    case CompressionMethodByte::FOR:
+        return getStaticCodec<CompressionCodecFOR>(setting);
+    default:
+        throw Exception(
+            ErrorCodes::UNKNOWN_COMPRESSION_METHOD,
+            "Unknown compression method byte: {:02x}",
+            static_cast<UInt16>(setting.method_byte));
+    }
+}
+
+CompressionCodecPtr CompressionCodecFactory::create(const CompressionSettings & settings)
+{
+    RUNTIME_CHECK(!settings.settings.empty());
+    CompressionCodecPtr codec = (settings.settings.size() > 1)
+        ? std::make_unique<CompressionCodecMultiple>(createCodecs(settings))
+        : create(settings.settings.front());
+    RUNTIME_CHECK(codec);
+#ifndef DBMS_PUBLIC_GTEST
+    RUNTIME_CHECK(codec->isCompression());
+#endif
+    return codec;
+}
+
+CompressionCodecPtr CompressionCodecFactory::createForDecompress(UInt8 method_byte)
+{
+    CompressionSetting setting(static_cast<CompressionMethodByte>(method_byte));
+    setting.data_type = CompressionDataType::Int8;
+    return create</*IS_DECOMPRESS*/ true>(setting);
+}
+
+Codecs CompressionCodecFactory::createCodecs(const CompressionSettings & settings)
+{
+    Codecs codecs;
+    codecs.reserve(settings.settings.size());
+    for (const auto & setting : settings.settings)
+    {
+        if (auto codec = create(setting); codec)
+            codecs.push_back(std::move(codec));
+    }
+    return codecs;
```

Comment:
Should we add a check that we will return at least one valid `codec`?

Reply:
Theoretically, CompressionCodecFactory::createCodecs can return a `vector<Codec>` with no element and `CompressionCodecMultiple` accept it.
https://github.com/pingcap/tiflash/blob/80deeaec1f3a109e8d830ce31b2d51ede095105a/dbms/src/IO/Compression/CompressionCodecMultiple.cpp#L32-L34
And it can pass the check you added here because it is a non-nullptr `CompressionCodecMultiple` instance.
https://github.com/pingcap/tiflash/blob/0e46c2e7f3b29d40d73458e760a453799146739c/dbms/src/IO/Compression/CompressionCodecFactory.cpp#L182-L193

I think we need to add an `assert(!codecs.empty())` in CompressionCodecFactory::createCodecs or the ctor of `CompressionCodecMultiple`

Reply:
done